homeassistant/components/mqtt/device_trigger.py

```diff
@@ -66,23 +66,53 @@
 )
 
 DEVICE_TRIGGERS = "mqtt_device_triggers"
-ATTACHED_DEVICE_TRIGGERS = "mqtt_attached_device_triggers"
+
+
+@attr.s(slots=True)
+class AttachedTrigger:
+    """Attached trigger settings."""
+
+    action = attr.ib(type=AutomationActionType)
+    automation_info = attr.ib(type=dict)
+    remove = attr.ib(type=CALLBACK_TYPE)
 
 
 @attr.s(slots=True)
 class Trigger:
     """Device trigger settings."""
 
+    attached_triggers = attr.ib(type=[AttachedTrigger])
     device_id = attr.ib(type=str)
-    discovery_id = attr.ib(type=str)
     hass = attr.ib(type=HomeAssistantType)
     payload = attr.ib(type=str)
     qos = attr.ib(type=int)
     subtype = attr.ib(type=str)
     topic = attr.ib(type=str)
     type = attr.ib(type=str)
 
-    async def attach_trigger(self, action, automation_info):
+    async def add_trigger(self, action, automation_info):
+        """Add MQTT trigger."""
+        attached_trigger = AttachedTrigger(action, automation_info, None)
+        self.attached_triggers.append(attached_trigger)
+
+        if self.topic is not None:
+            # If we know about the trigger, subscribe to MQTT topic
+            attached_trigger.remove = await self.attach_trigger(attached_trigger)
+
+        @callback
+        def async_remove() -> None:
+            """Remove trigger."""
+            if attached_trigger not in self.attached_triggers:
+                raise HomeAssistantError("Can't remove trigger twice")
+
+            index = self.attached_triggers.index(attached_trigger)
+            if self.attached_triggers[index].remove:
```

Comment:
Why this? Can't you just call 
```suggestion
            if attached_trigger.remove:
```